File: go/api/System.go
package api

import (
	"github.com/lvcdy/huawei-lte-api-go/session"
)

// System 对应 System.py。
type System struct {
	*session.ApiGroup
}

// NewSystem 创建 System API 分组。
func NewSystem(s *session.Session) *System {
	return &System{ApiGroup: session.NewApiGroup(s)}
}

// Devcapacity 对应 system/devcapacity。
func (s *System) Devcapacity() (map[string]interface{}, error) {
	return s.S.Get("system/devcapacity", nil, "api")
}

// Deviceinfo 对应 system/deviceinfo。
func (s *System) Deviceinfo() (map[string]interface{}, error) {
	return s.S.Get("system/deviceinfo", nil, "api")
}

// Deviceinfoex 对应 system/deviceinfoex。
func (s *System) Deviceinfoex() (map[string]interface{}, error) {
	return s.S.Get("system/deviceinfoex", nil, "api")
}

// Onlineupg 检查在线升级。对应 system/onlineupg (post_get, is_json)。
func (s *System) Onlineupg() (map[string]interface{}, error) {
	return s.S.PostGet("system/onlineupg", map[string]interface{}{
		"action": "check",
		"data": map[string]interface{}{
			"UpdateAction": 1,
		},
	}, false, "api", false, true)
}

// Onlinestate 获取在线状态。对应 system/onlinestate。
// Python 返回 list；Go 中返回的 map 需调用方自行转 []interface{}。
func (s *System) Onlinestate(devid string) (map[string]interface{}, error) {
	return s.S.Get("system/onlinestate", map[string]string{"devid": devid}, "api")
}

// Hostinfo 对应 system/HostInfo。
// Python 返回 list；Go 中返回的 map 需调用方自行转 []interface{}。
func (s *System) Hostinfo() (map[string]interface{}, error) {
	return s.S.Get("system/HostInfo", nil, "api")
}

File: src/api/system.rs
//! System API group (`api/System.py`).

use serde_json::Value;

use crate::connection::Connection;
use crate::errors::Result;
use crate::xml::{XmlMap, XmlValue};

use super::{get_value, post_get_value};

/// System API group.
pub struct System<'a> {
    conn: &'a Connection,
}

impl<'a> System<'a> {
    /// Build a group bound to `conn`.
    pub fn new(conn: &'a Connection) -> Self {
        System { conn }
    }

    /// `system/devcapacity`. Device capacity.
    pub fn devcapacity(&self) -> Result<Value> {
        get_value(self.conn, "api/system/devcapacity")
    }

    /// `system/deviceinfo`. Device information.
    pub fn deviceinfo(&self) -> Result<Value> {
        get_value(self.conn, "api/system/deviceinfo")
    }

    /// `system/deviceinfoex`. Extended device information.
    pub fn deviceinfoex(&self) -> Result<Value> {
        get_value(self.conn, "api/system/deviceinfoex")
    }

    /// Trigger an online upgrade check.
    pub fn onlineupg(&self) -> Result<Value> {
        let mut inner = XmlMap::new();
        inner.insert("UpdateAction".into(), XmlValue::Text("1".to_string()));
        let mut body = XmlMap::new();
        body.insert("action".into(), XmlValue::Text("check".to_string()));
        body.insert("data".into(), XmlValue::Map(inner));
        post_get_value(self.conn, "api/system/onlineupg", &body)
    }

    /// `system/onlinestate`. Online state of a device.
    pub fn onlinestate(&self, _devid: &str) -> Result<Value> {
        get_value(self.conn, "api/system/onlinestate")
    }

    /// `system/HostInfo`. Host information.
    pub fn hostinfo(&self) -> Result<Value> {
        get_value(self.conn, "api/system/HostInfo")
    }
}
